components/ws-manager-bridge/src/app-cluster-instance-controller.ts

```diff
@@ -0,0 +1,86 @@
+
+/**
+ * The WorkspaceInstance lifecycle is split between application clusters and workspace clusters on the transition from
+ * pending/building -> starting (cmp. WorkspacePhases here:
+ * https://github.com/gitpod-io/gitpod/blob/008ea3fadc89d4817cf3effc8a5b30eaf469fb1c/components/gitpod-protocol/src/workspace-instance.ts#L111).
+ *
+ * Before the transition, WorkspaceInstances belong to the respective app cluster, denoted by "instance.region === 'eu02'", for exmaple.
+ * After a WorkspaceInstance has been moved over to a workspace cluster, that moved "ownership" is reflected in said field.
+ * We maintain a constant connection (called "bridge") to all workspace clusters to be able to keep reality (workspace
+ * side) in sync with what we have in our DB/forward to clients.
+ *
+ * This class is meant to take the same responsibility for all WorkspaceInstances that have not (yet) been passed over
+ * to a workspace cluster for whatever reason. Here's a list of examples, prefixed by phase:
+ *  - "preparing": failed cleanup after failed call to wsManager.StartWorkspace
+ *  - "building": failed cleanup after failed image-build (which is still controlled by the application cluster,
+ *     although that might change in the future)
+ */
+@injectable()
+export class AppClusterWorkspaceInstancesController implements Disposable {
+    @inject(Configuration) protected readonly config: Configuration;
+
+    @inject(WorkspaceDB) protected readonly workspaceDb: WorkspaceDB;
+
+    @inject(WorkspaceInstanceController) protected readonly workspaceInstanceController: WorkspaceInstanceController;
+
+    protected readonly dispoables = new DisposableCollection();
+
+    public async start() {
+        const disposable = repeat(
+            async () => this.controlAppClusterManagedWorkspaceInstances(),
+            this.config.controllerIntervalSeconds * 1000,
+        );
+        this.dispoables.push(disposable);
+    }
+
+    protected async controlAppClusterManagedWorkspaceInstances() {
+        const appClusterInstallation = this.config.installation;
+
+        const span = TraceContext.startSpan("controlAppClusterManagedWorkspaceInstances");
+        const ctx = { span };
+        try {
+            log.info("Controlling app cluster instances", { installation: appClusterInstallation });
+
+            const notStoppedInstances = await this.workspaceDb.findRunningInstancesWithWorkspaces(
+                appClusterInstallation,
+                undefined,
+                false,
+            );
+            await this.workspaceInstanceController.controlNotStoppedAppClusterManagedInstanceTimeouts(
+                ctx,
+                notStoppedInstances,
+                appClusterInstallation,
+            );
+
+            log.info("Done controlling app cluster instances", {
+                installation: appClusterInstallation,
+                instancesCount: notStoppedInstances.length,
+            });
```

Comment:
Nice. This would be the perfect place to put any metrics which report duration & outcome. Shame we don't have metrics configured for `ws-manager-bridge`

Reply:
We could "cheat" and attach the duration of the run (and any outcome) to the log fields.